### PR TITLE
Refactor of frontend: auth flow, useeffects, loading data, routing... 

### PR DESCRIPTION
**Pull request by lin8231**
* There was a problem where routing wasn't working together with auth flow. This is because many things were kinda wonky. I will try to explain this to the best of my ability... 
   * Inner routing was completely broken: there was no way to route between pages once you're logged in, and you can't go to any page other than the default redirect page by changing the url.. This was because:
   * In App.js the user's auth state is checked inside a useEffect
   * Data is fetched in a useEffect 
   * The page route is defined inside a let pageRoute... this is not inside useEffect
   * The page route ternary statement executes before the auth state check - and by default, the user is NOT signed in (until we check and find out that the user is signed in). So the ternary statement sees that the user is not signed in, and redirects the user to /signin no matter what the old url was. Then in a while, the app detects that the user is signed in, and then the ternary statement executes again - but now the url is /signin which is not any of the signed in routes defined, hence it then redirects to whatever is in Redirect
   * To solve this problem, an authStateReceived state is made, this prevents the ternary statement from trying to load data and load routes before auth is successful. It checks that if authStateReceived is false, App.js shortcircuits and returns (nothing) before the routes could be defined. In the JSX body, we check if it is false, and if it's false, we show "signing in". 
   *  The useEffect which loads all the data and changes the loading state (of the data) now has authStateReceived as a dependency, so that when it changes, it reruns. 
       * setLoading(true) has now been pulled out of the promise.all because that's baaad (because thigns in Promise.all execute simultaneously, and we want setLoading to be true before any of the fetches even start executing. Before we were simply lucky that things work because setLoading happens to be the function that finishes the fastest. 
       *  FetchAndUpdateCompetitionInfo function now doesn't directly update the state as soon as it receives data, because that is bad practice for what we want to do. Instead this has now been refactored to be fetchCompetitionInfo and it doesn't just update the state and instead it returns an object that needs to be inserted into the state in a stateKey: value pair. After promise.all finishes executing this will return an array of the results of the promises, which we have named resolvedData. We then extract these objects from the array and set it to be a variable newCompInfo, we then set the state competitionInfo to be the newCompInfo.
     * When authStateDefined exists, short circuiting will not happen and routes will be defined - so now we add another check to see if competitionInfo has content on top of loading, to see if we are ready to load inner page content yet
* There is currently a makeshift thing going on which will be expanded in the next PR, where even if investments is not fetched (gets undefined), the app doesn't break, instead it just (currently) shows the sgonks list as "undefined". This feature will be expanded in a coming PR to make this prettier and more user friendly (aka not just an ugly "undefined") and also extend this logic to other sections (for if/when they fail to get fetched)
* Fetching user's id after auth is no longer done in App, it is done within the auth itself (id still hardcoded)

_**There is a lot of stuff in here, and I might have missed some stuff, please talk to me if you need further explanation on anything**_ 

# TLDR : a lot of bad things before, and as a result inner routing didn't work. inner routing works now and as a bonus the app doesn't break if investment doesn't exist

### CODE REVIEW

**Review by Phoebe-Khokgawe — APPROVED** [review]

File: project/frontend/src/App.js [hunk 1]
```diff
@@ -36,57 +41,57 @@ function App() {
...
+      };
+    }
   };
 
   React.useEffect(() => {
+    if (!authStateReceived) {
+      return;
+    }
+
     if (user.signedIn && user.id && compId) {
```
Comment (texm) [inline project/frontend/src/App.js:63]: While we're refactoring - I think this logic should be moved into a named function to better express what we're testing for here.
Also I think we should be more explicit about what is falsy - coercing integers into booleans can lead to some annoying errors (in general, not specifically here).

File: project/frontend/src/App.js [hunk 2]
```diff
@@ -36,57 +41,57 @@ function App() {
...
-          "./investments?user=" + user.id + "&competition=" + compId,
-          "investments"
-        ),
-        fetchAndUpdateCompetitionInfo("./trending", "trending"),
-      ])
-        .then(() => {
+      setLoading(true),
+        Promise.all([
+          fetchCompetitionInfo(
+            "./competitionInfo?user=" + user.id + "&competition=" + compId,
```
Comment (texm) [inline project/frontend/src/App.js]: Prefer to use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) for these

File: project/frontend/src/App.js [hunk 3]
```diff
@@ -36,57 +41,57 @@ function App() {
...
+            };
+          }
+
+          setCompetitionInfo(newCompInfo);
         });
     }
-  }, [user.id]);
+  }, [user.id, authStateReceived]);
 
   let pageRoute = !user.signedIn ? (
```
Comment (texm) [inline project/frontend/src/App.js:96]: I'd like to move away from using ternaries that are over 1 or 2 short lines as I think it doesn't convey intent as clearly.
I think initialising to a default value and then testing conditions by priority is more readable - brief example
```js
let pageContent = <div>Signing in...</div>

if (authStateReceived && !user.SignedIn) {
  pageContent = (
    <Layout>
      // ...
```

This logic is also long enough that it could be extracted into another function, perhaps this should be a child component?

File: project/frontend/src/App.js [hunk 4]
```diff
@@ -123,95 +128,14 @@ function App() {
       ></Route>
       <Route
         path="/competition"
-        render={() => (
-          <Competition
-          ></Competition>
-        )}
+        render={() => <Competition></Competition>}
```
Comment (texm) [inline project/frontend/src/App.js:131]: If we're not passing data to elements, may as well just write it as 
```js
<Competition />
```

There's a few other instances where this could be done too

File: project/frontend/src/App.js [hunk 5]
```diff
@@ -241,7 +164,11 @@ function App() {
             innerNav={compId != 0}
             compIdChanged={setCompId}
           ></HeaderBar>
-          <Layout>{pageRoute}</Layout>
+          {!authStateReceived ? (
+            <div>Signing in...</div>
+          ) : (
+            <Layout>{pageRoute}</Layout>
+          )}
```
Comment (texm) [inline project/frontend/src/App.js:171]: Continuation of above comment - this could be replaced with just a `{pageContent}`

File: project/frontend/src/components/SGonksLists/LongSGonksList/LongSGonksList.js [hunk 6]
```diff
@@ -3,18 +3,22 @@ import classes from "./LongSGonksList.module.css";
...
-        //TODO replace key with investment id once it's ready
-        key={i}
-        searchTerm={investment.searchItem}
-        buyInDate={investment.dateInvestedMilliSeconds}
-        amountInvested={investment.amtInvested}
-        currentValue={investment.currentValue}
-      ></LongSGonkRow>
-    );
-  });
+  const investmentRows = !props.investments ? (
```
Comment (texm) [inline project/frontend/src/components/SGonksLists/LongSGonksList/LongSGonksList.js:6]: (I know this file is in flux, feel free to disregard this comment)

I think again ternaries aren't helping readability here - would be clearer to use
```js
let investmentRows = <div>undefined</div>
if (props.investments !== undefined) {
  investmentRows = props.investments.map((investment, i) => {
    // ...
}
```

File: project/frontend/src/services/firebase.js [hunk 7]
```diff
@@ -35,6 +35,7 @@ export const onAuthStateChange = (callback) => {
         signedIn: true,
         name: user.displayName,
         email: user.email,
+        id: 1,
```
Comment (texm) [inline project/frontend/src/services/firebase.js:38]: We can't expect to receive an id here from firebase, as `id` is a concept/key that is in reference to our application database. Ideally we receive it in the response to our `POST` request to `"/authentication"` above.

Even more ideally - we shouldn't need our id to make calls to the backend after we are authenticated, as the backend should know who we are and what our id is. 

(IIRC there's also some bad practice stuff around exposing DB ids to clients, hence why [UUIDs](https://en.wikipedia.org/wiki/Universally_unique_identifier) are often used instead)

File: project/frontend/src/App.js (hunk 2 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Done.

File: project/frontend/src/App.js (hunk 4 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Done.

File: project/frontend/src/App.js (hunk 1 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Done :D

File: project/frontend/src/App.js (hunk 3 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Refactored :D

File: project/frontend/src/App.js (hunk 5 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Done.

File: project/frontend/src/services/firebase.js (hunk 7 above)
Comment (lin8231) [reply project/frontend/src/services/firebase.js:38]: Yeah I'm definitely aware of this. This is just the ID being hardcoded for now, when in reality it would be fetched from the backend. The reason it's hardcoded to 1 atm is because the last time I checked my own user (id=3) doesn't really have much data so there's nothing to display which is not very helpful for development. This will be fixed in issue #62

File: project/frontend/src/components/SGonksLists/LongSGonksList/LongSGonksList.js (hunk 6 above)
Comment (lin8231) [reply project/frontend/src/components/SGonksLists/LongSGonksList/LongSGonksList.js:6]: I will add a TODO comment in there to fix in the PR which deals with displaying partial data!

Comment (lin8231) [reply project/frontend/src/components/SGonksLists/LongSGonksList/LongSGonksList.js:6]: Done.

File: project/frontend/src/App.js [hunk 8]
```diff
@@ -1,236 +1,92 @@
...
-        setCompetitionInfo((prevState) => {
-          return {
-            ...prevState,
-            [stateKey]: data,
-          };
-        });
-      });
+  const isReadyForDataFetch = () => {
+    if (!user.signedIn) return false;
+    if (user.id === undefined) return false;
```
Comment (texm) [inline project/frontend/src/App.js:51]: Maybe we should check if `user.id` is 0, since we don't expect that either

File: project/frontend/src/hoc/PageRouter/PageRouter.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import { Route, Switch, Redirect } from "react-router-dom";
+
+import LandingPage from "../../containers/LandingPage/LandingPage";
+import SelectCompetition from "../../containers/SelectCompetition/SelectCompetition";
+import Dashboard from "../../containers/Dashboard/Dashboard";
+import MySGonks from "../../containers/MySGonks/MySGonks";
+import Marketplace from "../../containers/Marketplace/Marketplace";
+import Competition from "../../containers/Competition/Competition";
+
+import { NO_COMPETITION } from "../../App";
```
Comment (texm) [inline project/frontend/src/hoc/PageRouter/PageRouter.js:11]: may as well just define that in this file now

File: project/frontend/src/App.js (hunk 8 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Done.

File: project/frontend/src/App.js [hunk 10]
```diff
@@ -240,8 +96,21 @@ function App() {
...
+              <PageRouter
+                signedIn={user.signedIn}
+                compId={compId}
+                loading={loading}
+                compIdChanged={setCompId}
+                competitionInfo={competitionInfo}
+                updateCompId={setCompId}
+              />
+            </Layout>
+          )}
```
Comment (texm) [inline project/frontend/src/App.js:113]: This is better, I'd still like all of the logic to be handled by `PageRouter` though, eg end state looks like:
```js
<div className="App">
  <AuthContext>
    <HeaderBar
      loggedIn={user.signedIn}
      innerNav={compId != 0}
      compIdChanged={setCompId}
    />
    <PageRouter
      authStateReceived={authStateReceived}
      signedIn={user.signedIn}
      compId={compId}
      loading={loading}
      compIdChanged={setCompId}
      competitionInfo={competitionInfo}
      updateCompId={setCompId}
    />
  </AuthContext>
</div>
```

File: project/frontend/src/services/firebase.js (hunk 7 above)
Comment (texm) [reply project/frontend/src/services/firebase.js:38]: It's worth discussion about why we might want to change it, but probably not worth actually doing currently.

My concern around returning the id here is building logic elsewhere that assumes it will be available at the time that callback is called, when that won't be the case in future. Really we're not 100% authenticated until the backend has confirmed our identity as well, so the callback should be after that happens.

**Review by texm — APPROVED** [review]

File: project/frontend/src/hoc/PageRouter/PageRouter.js (hunk 9 above)
Comment (lin8231) [reply project/frontend/src/hoc/PageRouter/PageRouter.js:11]: I figured maybe it's better to have it just defined in one place to keep it consistent? That way if we ever need to change NO_COMPETITION's value to something else for whatever reason, we only need to check one place, rather than having to change it in multiple places. WDYT?

File: project/frontend/src/App.js (hunk 10 above)
Comment (lin8231) [reply project/frontend/src/App.js]: Refactored authStateReceived to be inside PageRouter as well, but I want to keep Layout on the outside, so that I don't have to wrap it around every return statement in PageRouter which would make it more verbose. Would this be ok? 
It now looks like:

```jsx
        <AuthContext>
          <HeaderBar
            loggedIn={user.signedIn}
            innerNav={compId != 0}
            compIdChanged={setCompId}
          />
          <Layout>
            <PageRouter
              authStateReceived={authStateReceived}
              signedIn={user.signedIn}
              compId={compId}
              loading={loading}
              competitionInfo={competitionInfo}
              updateCompId={setCompId}
            />
          </Layout>
        </AuthContext>

File: project/frontend/src/hoc/PageRouter/PageRouter.js (hunk 9 above)
Comment (texm) [reply project/frontend/src/hoc/PageRouter/PageRouter.js:11]: If we were using it throughout the codebase (or plan to use it in another component later on) it would be worthwhile to keep it in the top scoped file, however since the only use of it is in PageRouter I think we should keep it local to that.

File: project/frontend/src/services/firebase.js (hunk 7 above)
Comment (lin8231) [reply project/frontend/src/services/firebase.js:38]: Haven't really thought very deeply about this yet, but at a glance I was thinking we fetch here? 
Screen Shot 2021-02-08 at 9 20 50 pm
And use the call back in the .then()?

File: project/frontend/src/hoc/PageRouter/PageRouter.js (hunk 9 above)
Comment (texm) [reply project/frontend/src/hoc/PageRouter/PageRouter.js:11]: Disregard, just saw that we're using it in App.js now 🤦

File: project/frontend/src/App.js (hunk 10 above)
Comment (texm) [reply project/frontend/src/App.js]: Makes sense, nice one :)

File: project/frontend/src/services/firebase.js (hunk 7 above)
Comment (texm) [reply project/frontend/src/services/firebase.js:38]: Re-reading `services/firebase.js`, I think we should combine `signInWithGoogle` and `onAuthStateChange` into something like this:

```js
const signInWithGoogle = (callback) => {
  firebase.auth()
  .signInWithPopup(provider)
  .then(result => {
    let credential = result.credential;
    let token = credential.accessToken;

    let user = result.user;
    
    user.getIdToken(true)
      .then(tokenResult => {
        // POST idToken to /authentication, get id from response
        // call callback with user details and id
        // logged in!
      });
  }); 
}
```

WDYT?

Comment (texm) [reply project/frontend/src/services/firebase.js:38]: Refer to docs https://firebase.google.com/docs/auth/admin/manage-cookies#sign_in

Comment (lin8231) [reply project/frontend/src/services/firebase.js:38]: I tried this out quickly, and I seem to fall into an infinite loop of signing in and I'm not entirely sure why. Maybe @Phoebe-Khokgawe might understand this better, could you have a look at this?
